code/modules/projectiles/projectile.dm

```diff
@@ -674,268 +735,412 @@
+	var/atom/handle_atom = last_impact_turf || src
+	var/datum/point/new_point = impact ? new /datum/point(handle_atom.x, handle_atom.y, handle_atom.z, impact_x, impact_y) : RETURN_PRECISE_POINT(tile_center ? loc : src)
+	if (!broken_segment)
+		beam_points[last_point] = new_point
+	beam_points[new_point] = null
+	last_point = new_point
+
+/obj/projectile/forceMove(atom/target)
+	if (hitscan && !isnull(beam_points))
+		create_hitscan_point()
+	. = ..()
+	if(!isturf(loc) || !isturf(target) || !z || QDELETED(src) || deletion_queued)
+		return .
+	if (!hitscan || isnull(movement_vector) || isnull(beam_points) || free_hitscan_forceMove)
+		return .
+	// Create firing VFX and start a new chain because we most likely got teleported
+	generate_hitscan_tracers(impact = FALSE)
+	original_angle = angle
+	record_hitscan_start(offset = FALSE)
+
+/obj/projectile/proc/generate_hitscan_tracers(impact = TRUE)
+	if (!length(beam_points))
+		return
+
+	if (impact)
+		create_hitscan_point(impact = TRUE)
+
+	if (tracer_type)
+		for (var/beam_point in beam_points)
+			generate_tracer(beam_point)
+
+	if (muzzle_type)
+		var/datum/point/start_point = beam_points[1]
+		var/atom/movable/muzzle_effect = new muzzle_type(loc)
+		start_point.move_atom_to_src(muzzle_effect)
+		var/matrix/matrix = new
+		matrix.Turn(original_angle)
+		muzzle_effect.transform = matrix
+		muzzle_effect.color =  color
+		muzzle_effect.set_light(muzzle_flash_range, muzzle_flash_intensity, muzzle_flash_color_override || color)
+		QDEL_IN(muzzle_effect, PROJECTILE_TRACER_DURATION)
+
+	if (impact_type)
+		var/atom/movable/impact_effect = new impact_type(loc)
+		last_point.move_atom_to_src(impact_effect)
+		var/matrix/matrix = new
+		matrix.Turn(angle)
+		impact_effect.transform = matrix
+		impact_effect.color =  color
+		impact_effect.set_light(impact_light_range, impact_light_intensity, impact_light_color_override || color)
+		QDEL_IN(impact_effect, PROJECTILE_TRACER_DURATION)
+
+/obj/projectile/proc/generate_tracer(datum/point/start_point)
+	if (isnull(beam_points[start_point]))
+		return
+
+	var/datum/point/end_point = beam_points[start_point]
+	var/datum/point/midpoint = point_midpoint_points(start_point, end_point)
+	var/obj/effect/projectile/tracer/tracer_effect = new tracer_type(midpoint.return_turf())
+	tracer_effect.apply_vars(angle_between_points(start_point, end_point), midpoint.pixel_x, midpoint.pixel_y, color, pixel_length_between_points(start_point, end_point) / ICON_SIZE_ALL)
+	QDEL_IN(tracer_effect, PROJECTILE_TRACER_DURATION)
+
+	if (!hitscan_light_range || !hitscan_light_intensity)
+		return
+
+	var/self_ref = REF(src)
+	var/list/turf/light_line = get_line(start_point.return_turf(), end_point.return_turf())
+	for (var/turf/light_turf as anything in light_line)
+		var/located_effect = FALSE
+		for(var/obj/effect/abstract/projectile_lighting/light_effect in light_turf)
+			if(light_effect.owner == self_ref)
+				located_effect = TRUE
+				break
```

Comment:
is this really the best way of doing this?

Reply:
can't think of anything much better, we'll have to iterate through all of em anyways

Reply:
depending on how often this is done it might be worth it to use an assoc list (I don't have a good feel for when this proc is called tho)